### ISSUE TITLE
= ANY does not return any rows

### ISSUE THREAD

**Issue by Lewenhaupt** [oguimbal/pg-mem#338]
# Describe the bug

Does not appear as if ANY queries with ARRAYs are supported. Tried to query on just one id and then it works, but as soon as I attempt to query using ANY and ARRAY (i.e. ANY(ARRAY[1,2]) ) it does not return anything.

# To Reproduce

```sql
-- create tables
CREATE TABLE "user" ("id" SERIAL NOT NULL, "name" text NOT NULL, CONSTRAINT "PK_cace4a159ff9f2512dd42373760" PRIMARY KEY ("id"));

-- insert data
INSERT INTO "user"("name") VALUES ('me') RETURNING "id";
SELECT "id" FROM "user" WHERE "id"=1; -- Works and returns the row
SELECT "id" FROM "user" WHERE "id"=ANY(ARRAY[1]); -- Does not work
SELECT "id" FROM "user" WHERE "id"=ANY('{1}'); -- Does not work either

```

# pg-mem version
2.6.13

**Comment by Lewenhaupt** [comment]
@oguimbal 
Yeah I can't get it to work when actually running it in the playground or in my own tests.
But I did try to add a unit test to pg-mem to see if that had the same problem, but it didn't?

Added this in select.queries.spec.ts and it passes...
`it('can filter on any', () => {
        stuff();
        expect(many(`select * from test where val = any('{999}');`))
            .to.deep.equal([
                { txt: 'A', val: 999 },
            ])
    });`

**Comment by Lewenhaupt** [comment]
@oguimbal Any input?

**Comment by rswheeldon** [comment]
Interestingly if you do:
```
select * from "user" where "name" = any(array['me']);
```
on your example it *does* work. What the difference is I haven't yet figured out but it's annoying because the thing I'm trying to test runs into similar problems.

**Comment by rswheeldon** [comment]
@Lewenhaupt Interestingly this also works - which might be a useful workaround for your case?
```
select * from "user" where array["id"::integer] && array[1];
```

**Comment by earfman** [comment]
Found the root cause — and it explains the mystery in this thread.

**The variable is the index.** `col = ANY(array)` fails only when the column is indexed; it works on non-indexed columns:

- @Lewenhaupt's repro fails because `"id"` is a `SERIAL PRIMARY KEY` (indexed).
- @rswheeldon's `"name" = any(array['me'])` works because `name` has no index.
- The unit test @Lewenhaupt added passed because the test table's `val` column has no index either — so the test never exercised the broken path.

Minimal demonstration:

```js
const { newDb } = require('pg-mem');
const db = newDb();
db.public.many(`create table t (id text primary key);
                insert into t values ('A'), ('B'), ('C');`);
db.public.many(`select * from t where id = any(array['A','C'])`); // -> []  (bug)
// same table WITHOUT the primary key -> [A, C] correctly
```

**Why:** in `src/transforms/build-filter.ts`, `buildComparison` optimizes `col = <constant>` into a scalar index equality (`EqFilter`). When the operand is `ANY(array)`, the constant is the *whole array*, so it performs an index lookup for a key equal to `['A','C']` — which matches no row. Without an index that path is never taken, and the seq-scan fallback evaluates `ANY` element-wise, correctly. (The same broken path also affects `> ANY` and `!= ANY` on indexed columns.)

I've opened a PR with a minimal fix (skip the scalar index optimization when an operand is `ANY`/`ALL`, falling back to the seq scan) plus a regression test on an indexed column — the missing sibling of the existing ANY tests: #477

**Comment by Lewenhaupt** [comment]
Oh, very nice find! 😃
